lib/src/rules/conformance/banned_property_write.dart

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../../analyzer.dart';
+import 'conformance_rule.dart';
+import 'interop_helpers.dart';
+import 'web_bindings.dart';
+
+class _BannedPropertyWriteCode extends SecurityLintCode {
+  _BannedPropertyWriteCode(
+      {required String name,
+      required String problemMessage,
+      required String correctionMessage})
+      : super(name, problemMessage, correctionMessage: correctionMessage);
+}
+
+class BannedPropertyWrite extends ConformanceRule {
+  final String? nativeType;
+  final String? nativeProperty;
+  final String? dartHtmlType;
+  final String? dartHtmlProperty;
+  BannedPropertyWrite(
+      {required String name,
+      required String description,
+      required String details,
+      required this.nativeType,
+      required this.nativeProperty})
+      : dartHtmlType = null,
+        dartHtmlProperty = null,
```

Comment:
I like the idea of using some form of a union type to express the disjointedness. I can imagine the two fields scenario (although one of them will still be `null`).

I'm a little confused, however, at the polymorphic proposal. One of the benefits of using a `null` value is that equality checks against it e.g. `dartHtmlProperty` can check two things at once: both whether the rule is using the native binding and whether the value matches. For example `nativeProperty == rule.nativeProperty` (in the case where `nativeProperty` is non-null) tells me both of these things. With the two fields case, I can imagine it being `nativeProperty == rule.nativeDescriptor?.property` which will operate similarly.

I'd imagine the polymorphic case would have to place this behavior into several methods on the abstract class e.g.

```
abstract class PropertyDescriptor {
  String get property;
}

abstract class NativePropertyDescriptor ... {
  ...
  @override
  String get property => _nativeProperty;
}
```

and then the above equality check would become `descriptor is NativePropertyDescriptor && nativeProperty == descriptor.property`. It feels a bit wonky/excessive boilerplate to have `is` checks often (or a bool caching the result), but I might be interpreting your suggestion wrong. Thoughts?

Reply:
The polymorphism case might be too much, but I'd probably handle something like what you're describing (assuming I'm understanding it correctly, as
```dart
abstract class PropertyDescriptor {
  bool matches(String property);
}
class NativePropertyDescriptor implements PropertyDescriptor {
  ...
  @override
  bool matches(String property) => property == _nativeProperty;
}
```
Or maybe even something a little higher level like `void checkInvocation(MethodInvocation node)` that would just do all of the checking appropriate for the kind of property. Think of it as a dispatch table to property-specific behavior. Of course, that might force too much code to be duplicated; I haven't actually tried to rewrite the code that way.

Mostly I was thinking about ways to eliminate the need for nullable types.

Reply:
I ended up doing something like that except instead of `matches`, it's a getter (because the values themselves are needed for map lookups). It's not very pretty, but it does remove nullables out of the picture. The rule gets exactly one descriptor, and methods to determine if it's a `dart:html` or native descriptor.

We could have probably gotten rid of nullables by using the empty string, but this works better when this code moves to shared helper functions (for banned property gets, etc.). Instead of passing 4 fields, you pass the descriptor to a helper method.

I agree that adding classes to check nodes separately for the two types of descriptor will lead to a good amount of code-duplication.